### PR TITLE
pr_gateway refactoring

### PR DESCRIPTION
**Pull request by apahim**
https://issues.redhat.com/browse/APPSRE-2375

I'd higly reccomend reviewing it per-commit basis & reading the inner README as the first step :)

At the moment, this does nothing.

To use it as an SQS consumer, we have to change here:

https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/hack/periodic_job.sh#L49

From `gitlab-pr-submitter` to `gitlab-mr-sqs-consumer`.

Using it as SQS producer will be a later PR to this repo.

MR examples:

- https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/8760
- https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/8776
- https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/8789
- https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/8798

### CODE REVIEW

File: reconcile/gitlab_mr_sqs_consumer.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        for message in messages:
+            receipt_handle, body = message[0], message[1]
+
+            logging.info('received message %s with body %s',
+                         receipt_handle[:6], json.dumps(body))
+
+            if not dry_run:
+                merge_request = mr.init_from_sqs_message(body)
+                merge_request.submit_to_gitlab(gitlab_cli=gitlab_cli)
+                sqs_cli.delete_message(receipt_handle)
```
Comment (maorfr) [inline reconcile/gitlab_mr_sqs_consumer.py:45]: can we defer the delete_message? if a message is faulty i think that we currently don't want to keep it in the queue (we'll need to think about an error queue at a later point, but not in scope for this PR).

File: utils/mr/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+To send a Merge Request to SQS, create the corresponding MergeRequest object:
+
+```python
+from utils.mr import CreateAppInterfaceNotificator
+
+
+notification = {
+    'notification_type': 'Outage',
+    'description': 'The AppSRE team is current investigating an outage',
```
Comment (maorfr) [inline utils/mr/README.md:15]: typo: currently

File: utils/mr/base.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+        if gitlab_cli.mr_exists(title=self.title):
+            return
+
+        gitlab_cli.create_branch(new_branch=self.branch,
+                                 source_branch=self.main_branch)
+
+        try:
+            self.process(gitlab_cli=gitlab_cli)
+        except (CancelMergeRequest, GitlabError) as details:
```
Comment (maorfr) [inline utils/mr/base.py:119]: :heart:

File: utils/mr/cloud_ingress.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import ruamel.yaml as yaml
```
Comment (maorfr) [inline utils/mr/cloud_ingress.py:1]: this one is no longer needed. a different direction was chosen.

File: utils/mr/labels.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+AUTO_MERGE = 'automerge'
```
Comment (maorfr) [inline utils/mr/labels.py:1]: we'll probably need groups of labels here in some way (thinking about gitlab_housekeeping)

File: reconcile/gitlab_mr_sqs_consumer.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            break
+
+        for message in messages:
+            receipt_handle, body = message[0], message[1]
+
+            logging.info('received message %s with body %s',
+                         receipt_handle[:6], json.dumps(body))
+
+            if not dry_run:
+                merge_request = mr.init_from_sqs_message(body)
```
Comment (maorfr) [inline reconcile/gitlab_mr_sqs_consumer.py:43]: i know we are checking it in relevant MR types, but would it make sense to check that the merge_request is not empty (no changes)?

**Review by maorfr — APPROVED** [review]
LGTM

this is amazing work! i love the idea of `MergeRequestBase` and especially `process`.
i've added some minor comments.
the documentation is great! thank you!

File: utils/mr/labels.py (hunk 5 above)
Comment (apahim) [reply utils/mr/labels.py:1]: Yeah.. we can extend them when the time comes

File: reconcile/gitlab_mr_sqs_consumer.py (hunk 6 above)
Comment (apahim) [reply reconcile/gitlab_mr_sqs_consumer.py:43]: Good point.

**Pull request merged**